File: app/api/schemas/graph.py
"""
Pydantic response schemas for the graph API (UI2).

Condition/threshold shapes are genuinely variable across the ~60 test
classes (see rule_condition_parser.py) — using dict[str, Any] here
instead of forcing every possible field onto one rigid schema, same
flexible-JSON philosophy already used for rule_conditions.structured_data
on the Postgres side.
"""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel


class RuleGraphDetail(BaseModel):
    rule: dict[str, Any]
    references: list[dict[str, Any]]
    conditions: list[dict[str, Any]]
    log_sources: list[str]
    mitre: list[dict[str, Any]]
    followed_by: list[dict[str, Any]]


class BuildingBlockDependent(BaseModel):
    rule_id: int
    identifier: str | None
    name: str | None
    threshold: dict[str, Any]


class FieldSearchResult(BaseModel):
    rule_id: int
    identifier: str | None
    name: str | None
    operator: str | None
    values: list[str] | None


class TechniqueSearchResult(BaseModel):
    rule_id: int
    identifier: str | None
    name: str | None

File: frontend/src/lib/types/graph.ts
export interface FieldSearchResult {
	rule_id: number;
	identifier: string | null;
	name: string | null;
	operator: string | null;
	values: string[] | null;
}

export interface BuildingBlockDependent {
	rule_id: number;
	identifier: string | null;
	name: string | null;
	threshold: Record<string, unknown>;
}

export interface RuleGraphDetail {
	rule: {
		rule_id: number;
		identifier: string | null;
		name: string | null;
		object_type: string | null;
		enabled: boolean | null;
		type: string | null;
		owner: string | null;
		origin: string | null;
	};
	references: {
		identifier: string | null;
		name: string | null;
		threshold: Record<string, unknown>;
	}[];
	conditions: {
		test_class: string;
		negated: boolean;
		raw_text: string | null;
		field?: string | null;
		operator?: string | null;
		values?: string[] | null;
	}[];
	log_sources: string[];
	mitre: {
		technique_id: string | null;
		technique_name: string | null;
		tactic_id: string | null;
		tactic_name: string | null;
	}[];
	followed_by: {
		source_identifier: string | null;
		source_name: string | null;
		target_identifier: string | null;
		target_name: string | null;
		relationship: Record<string, unknown>;
	}[];
}